File: android/app/src/main/java/com/kurt/bgmate/data/repository/GameRepositoryImpl.kt
package com.kurt.bgmate.data.repository

import com.kurt.bgmate.data.local.BoardGameDao
import com.kurt.bgmate.data.local.BoardGameEntity
import com.kurt.bgmate.data.local.SessionDao
import com.kurt.bgmate.data.local.toDomain
import com.kurt.bgmate.data.local.toEntity
import com.kurt.bgmate.data.remote.BggRemoteDataSource
import com.kurt.bgmate.domain.model.BoardGame
import com.kurt.bgmate.domain.model.ScoreSession
import com.kurt.bgmate.domain.repository.GameRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import javax.inject.Inject

class GameRepositoryImpl @Inject constructor(
    private val gameDao: BoardGameDao,
    val sessionDao: SessionDao,
    private val bggRemoteDataSource: BggRemoteDataSource
) :
    GameRepository {
    private val games = mutableListOf<BoardGame>()

    override fun getGames(): List<BoardGame> = games.toList()

    override suspend fun addGame(name: String) {
        // 임시 ID 생성 — BGG 승인 후 실제 bggId로 교체 예정
        val tempId = "local_${System.currentTimeMillis()}"
        gameDao.insertGame(BoardGameEntity(bggId = tempId, name = name))
    }

    override suspend fun addGame(game: BoardGame) {
        gameDao.insertGame(game.toEntity())
    }

    override suspend fun removeGame(name: String) {
       val entity =  gameDao.observeGames().first().first() { it.name == name }
        gameDao.delete(entity)
    }

    override suspend fun updateGame(old: String, new: String) {
        val entity =  gameDao.observeGames().first().first() { it.name == old } ?: return
        gameDao.updateGameName(entity.bggId, new)
    }

    override suspend fun searchGames(query: String): List<BoardGame> {
        return bggRemoteDataSource.searchGames(query)
    }

    override suspend fun getGameById(id: String): BoardGame? {
        return gameDao.getGameById(id = id)?.toDomain()
    }

    override fun observeGames(): Flow<List<BoardGame>> = gameDao.observeGames().map { list -> list.map { it.toDomain() } }

    override fun observeSessionHistory(): Flow<List<ScoreSession>> =
        sessionDao.observeAllSessions().map { list -> list.map { it.toDomain() } }

    override suspend fun getSessionById(sessionId: Long): ScoreSession? =
        sessionDao.getSessionWithDetails(sessionId)?.toDomain()
}

File: flutter/lib/data/repository/game_repository_impl.dart
import 'package:bgmate_flutter/data/local/game_dao.dart';
import 'package:bgmate_flutter/data/local/game_table_mapper.dart';
import 'package:bgmate_flutter/data/remote/bgg_remote_data_source.dart';
import 'package:bgmate_flutter/domain/model/board_game.dart';
import 'package:bgmate_flutter/domain/repository/game_repository.dart';

class GameRepositoryImpl implements GameRepository {
  final GameDao _gameDao;
  final BggRemoteDataSource _bggRemoteDataSource;

  GameRepositoryImpl(this._gameDao, this._bggRemoteDataSource);

  @override
  Future<List<BoardGame>> getCollection() async {
    final entities = await _gameDao.getAll();
    return entities.map((e) => e.toDomain()).toList();
  }

  @override
  Future<List<BoardGame>> searchBgg(String query) async {
    return _bggRemoteDataSource.searchGames(query);
  }

  @override
  Future<List<BoardGame>> enrichWithDetails(List<BoardGame> games) async {
    if (games.isEmpty) return games;
    final ids = games.map((g) => g.bggId).toList();
    final details = await _bggRemoteDataSource.getThingDetails(ids);
    final detailMap = {for (final d in details) d.bggId: d};
    return games
        .map((g) => detailMap[g.bggId] ?? g)
        .toList();
  }

  @override
  Future<BoardGame?> getGame(int id) async {
    final data = await _gameDao.getGame(id);
    return data?.toDomain();
  }

  @override
  Future<void> addToCollection(BoardGame game) async {
    return await _gameDao.upsert(game.toCompanion());
  }

  @override
  Future<void> removeFromCollection(BoardGame game) async {
    return _gameDao.deleteByBggId(game.bggId);
  }

  @override
  Stream<List<BoardGame>> watchGames() {
    return _gameDao.watchAll().map(
      (entities) => entities.map((e) => e.toDomain()).toList(),
    );
  }
}
